### ISSUE TITLE
RPi: Fails to start on a raspberry pi

### ISSUE THREAD

**Issue by 44yu5h** [sugarlabs/browse-activity#134]
Browser activity fails to start on Raspberry Pi
---

- `v207` gets installed instead of `v208` after installing `sucrose` apt package. (I don't know whom to ping for apt)

  >` apt show sugar-browse-activity` returns
  > Package: sugar-browse-activity
  > Version: 207-2

  `v207` has rendering issues but we can ignore this and focus on `v208`:
  (will upload a screenshot soon)

- after updating, `v208` (latest as of now) returns:

  ```python
  Traceback (most recent call last): File "/usr/bin/sugar-activity3", line 5, in <module> activityinstance.main() File "/usr/lib/python3/dist-packages/sugar3/activity/activityinstance.py", line 179, in main module = import(module_name) ^^^^^^^^^^^^^^^^^^^^^^^ File "/home/pi/Activities/browse-activity/webactivity.py", line 31, in <module> gi.require_version('Soup', '3.0') File "/usr/lib/python3/dist-packages/gi/init.py", line 129, in require_version raise ValueError('Namespace %s not available for version %s' % ValueError: Namespace Soup not available for version 3.0
  ```

---
### To summarize, there are 2 issues:
 - Installing `sucrose` from apt, installs the old version (`v207`) which is "buggy"
 - `v208` returns an error stating `Soup 3.0` is missing

**Comment by 44yu5h** [comment]
The problem is the missing dependencies required in the `webactivity.py` and `browser.py`'s import statements

What all were missing from RPi OS that were "`gi.required`":
- `Soup 3.0` (`2.4` was present)
- `Webkit2 4.1` (`4.0` was present)

Installing these makes the browser run normally

---

 **Steps to fix:**

1. installing `Soup 3.0`:
      ```
     sudo apt install libsoup-3.0
      ```
    > Another error pops up stating Soup 2.4 is already loaded instead of Soup 3.0
    > (this was caused because `webkit2 4.1` was not available)
    > If you look at the code, it makes the `incompatile` variable True when `gi.require` fails but later imports it regardless. This imports `Webkit2 4.0` which automatically loads `Soup 2.4`
 3. Installing `Webkit2 4.1`:
      ```
      sudo apt install gir1.2-webkit2-4.1
      export GI_TYPELIB_PATH=/usr/lib/girepository-1.0
      ```

**Comment by 44yu5h** [comment]
I don't understand how to implement the fix. Maybe someone can suggest suggest or do the fix.\
Thanks

PS: Ways to determine if the device is RPi: 
1. `raspi-config nonint is_pi` returns nothing; returns command not found otherwsie
2. `raspi-config nonint get_pi_type` return a number;    " "
3. `cat /proc/device-tree/model` returns something like "Raspberry Pi 3 Model B Rev 1.2" (keyword: Raspberry Pi)

**Comment by quozl** [comment]
You cannot reasonably implement a fix here.  It is the responsibility of Raspberry Pi OS team to fix this.  You could work with them to get the fix done.

You haven't mentioned which version of Raspberry Pi OS you tested.  Is there a later version?  Is the package simply copied from Debian, in which case you need to arrange for the fix in Debian?  What is the Debian package version?

That should get you started.

**Comment by 44yu5h** [comment]
@quozl Thanks for replying. I'm sorry for not getting back to you sooner, I've been a little caught up with school stuff. 

> It is the responsibility of Raspberry Pi OS team to fix this. You could work with them to get the fix done.

I've sent them a message, waiting for a reply. 

> You haven't mentioned which version of Raspberry Pi OS you tested.

It's the latest one. Bookworm 12\
I was using the lite version though (the one we use with Sugar). Still, I don't think there should be an older version of Soup in the lite version. I'll verify it once anyway.

---

Replacing the required version to `Soup 2.4` and `Webkit2 4.0` does the trick.\
Will implementing a simple Raspberry Pi check and using the above version cause too much of a bloat?\
It'll be great if we can make the browser run out of the box since it's needed a lot with RPi

**Comment by quozl** [comment]
Okay, thanks.  Raspberry Pi OS packages for Sugar are built from the Debian packages.  Please use `dpkg-query -W sugar-browse-activity` to tell me the version.

Debian Bookworm is stable, which is software engineering technical term for "we won't be changing it unless there is something _really_ important to fix".  Stable doesn't mean works well.

Debian Trixie is the testing release.

https://packages.debian.org/search?keywords=sugar-browse-activity shows that Browse 207 is in Bookworm, and Browse 208 is in Trixie.

https://packages.debian.org/bookworm/sugar-browse-activity shows the dependencies in Bookworm include Soup 2.4.

https://packages.debian.org/sid/sugar-browse-activity shows the dependencies in Trixie include Soup 3.0.

https://metadata.ftp-master.debian.org/changelogs//main/s/sugar-browse-activity/sugar-browse-activity_208-1_changelog shows the changes that led to the current releases by Debian.

So there's nothing you need to do about this problem in Debian or Raspberry Pi OS except wait for the next release of Debian to be adopted by Raspberry Pi OS.

The real solution to this problem is to regularly test Sugar and activities using Debian Sid unstable, both on ordinary computers and also on Raspberry Pi.  That way problems that appear will be fixed more quickly, hopefully before Debian is released, and therefore before Raspberry Pi OS acquires the Debian packages.

Here's a graph of the flow;

```log
     ┌────────────────────────────────────────────┐
     │              raspberry pi os               │ ◀─────┐
     └────────────────────────────────────────────┘       │
       │                                                  │
       ▼                                                  │
     ┌────────────────────────────────────────────┐       │
     │                  enduser                   │       │
     └────────────────────────────────────────────┘       │
       │                                                  │
       ▼                                                  │
     ┌────────────────────────────────────────────┐       │
     │              bugs.debian.org               │ ─┐    │
     └────────────────────────────────────────────┘  │    │
       │                                             │    │
       ▼                                             │    │
     ┌────────────────────────────────────────────┐  │    │
  ┌▶ │                   issues                   │  │    │
  │  └────────────────────────────────────────────┘  │    │
  │    │                                             │    │
  │    ▼                                             │    │
  │  ┌────────────────────────────────────────────┐  │    │
  └─ │                 developer                  │  │    │
     └────────────────────────────────────────────┘  │    │
       │                                             │    │
       ▼                                             │    │
     ┌────────────────────────────────────────────┐  │    │
     │ github.com/sugarlabs/sugar-browse-activity │  │    │
     └────────────────────────────────────────────┘  │    │
       │                                             │    │
       ▼                                             │    │
     ┌────────────────────────────────────────────┐  │    │
     │                   debian                   │ ◀┘    │
     └────────────────────────────────────────────┘       │
       │                                                  │
       └──────────────────────────────────────────────────┘
```

(thanks to dot-to-ascii)

**Comment by quozl** [comment]
Also, addressing your wording of issues;

>  Installing sucrose from apt, installs the old version (v207) which is "buggy"

Yes, and will be fixed in the next Raspberry Pi OS release after Debian Trixie.  Raspberry Pi OS team could, if they wish, take the updated package from Debian Trixie and apply it to their derivative of Debian Bookworm.  While that would be nice, there's no clear justification for them to do this.

> v208 returns an error stating Soup 3.0 is missing

That happened because you installed v208 by hand rather than use the Debian Trixie package.  Our activities generally don't resolve their own dependencies, instead we guide packagers to do it for us, or we embed the dependency temporarily.

Looking at why Browse uses Soup;
* use of form_encode_hash to encode a search query parameter,
* making a cookie for authentication with the School Server, which is unlikely to be used now, and it seems the cookie jar is deleted without being used, so perhaps this code can go if it has no effect.

If we can rid ourselves of Soup, that would be great!  Perhaps that is less effort than alternatives.

**Comment by 44yu5h** [comment]
@quozl Thank you for explaining that in so much detail! That makes perfect sense now.

> That happened because you installed v208 by hand rather than use the Debian Trixie package.

Exactly! I always forget how packages manage dependencies. So updating the apt repo is all that's needed or wait for the next RPi OS release. 

> If we can rid ourselves of Soup, that would be great!

I don't know how that works so I will let it be for now.

---

Bottom line, I will make sure to include both the dependencies in the RPi image that we generate to make things work out of the box for now.

Thank you so much again!
